omegaconf/dictconfig.py

```diff
@@ -146,103 +145,103 @@ def _validate_get(self, key: Any, value: Any = None) -> None:
                     key=key, value=value, cause=ConfigAttributeError(msg)
                 )
 
-    def _validate_merge(self, key: Any, value: Any) -> None:
-        self._validate_set_merge_impl(key, value, is_assign=False)
-
     def _validate_set(self, key: Any, value: Any) -> None:
-        self._validate_set_merge_impl(key, value, is_assign=True)
-
-    def _validate_set_merge_impl(self, key: Any, value: Any, is_assign: bool) -> None:
         from omegaconf import OmegaConf
 
         vk = get_value_kind(value)
         if vk in (ValueKind.INTERPOLATION, ValueKind.STR_INTERPOLATION):
             return
-
-        if OmegaConf.is_none(value):
-            if key is not None:
-                child = self._get_node(key)
-                if child is not None and not child._is_optional():
-                    self._format_and_raise(
-                        key=key,
-                        value=value,
-                        cause=ValidationError("child '$FULL_KEY' is not Optional"),
-                    )
-            else:
-                if not self._is_optional():
-                    self._format_and_raise(
-                        key=None,
-                        value=value,
-                        cause=ValidationError("field '$FULL_KEY' is not Optional"),
-                    )
-
-        if value == "???":
+        self._validate_non_optional(key, value)
+        if value == "???" or value is None:
             return
 
-        if is_assign and isinstance(value, ValueNode) and self._has_element_type():
-            self._check_assign_value_node(key, value)
-            return
+        target = self._get_node(key) if key is not None else self
 
-        target: Optional[Node]
-        if key is None:
-            target = self
-        else:
-            target = self._get_node(key)
+        assign_value_node = isinstance(value, ValueNode) and self._has_element_type()
+        target_has_ref_type = isinstance(
+            target, DictConfig
+        ) and target._metadata.ref_type not in (Any, dict)
+        is_invalid_target = target is None or not target_has_ref_type
 
-        if target is None:
+        if is_invalid_target and not assign_value_node:
             return
 
-        def is_typed(c: Any) -> bool:
-            return isinstance(c, DictConfig) and c._metadata.ref_type not in (Any, dict)
+        target_type = (
+            self._metadata.element_type if assign_value_node else target._metadata.ref_type  # type: ignore
+        )
+        value_type = OmegaConf.get_type(value)
 
-        if not is_typed(target):
+        if is_generic_type(target_type):
             return
```

Comment:
This also reads really weird to me.
Why is ValueNode assignment getting special treatment, and what is that treatment exactly.
Also, under what circumstances is the element_type a generic type?
Is it a case like this one?
```
foo : Dict[str, List[int]]
```
If so, it seems completely illegal to assign a value node to it, so I don't understand why you return here.

Reply:
> This also reads really weird to me.
> Why is ValueNode assignment getting special treatment, and what is that treatment exactly.

ValueNode is getting special treatment because when assigning a structured config into a DictConfig it first converts structured config's data to Nodes as opposed to every other instance(raw values). 

> Also, under what circumstances is the element_type a generic type?
> Is it a case like this one?
> 
> ```
> foo : Dict[str, List[int]]
> ```
> 
> If so, it seems completely illegal to assign a value node to it, so I don't understand why you return here.

We return when is a generic container because it can cause errors like `omegaconf.errors.ConfigTypeError: Subscripted generics cannot be used with class and instance checks` on 
```python
validation_error = (
            target_type is not None
            and value_type is not None
            and not issubclass(value_type, target_type)
        )
```

Reply:
> ValueNode is getting special treatment because when assigning a structured config into a DictConfig it first converts structured config's data to Nodes as opposed to every other instance(raw values).

Read this sentence again.
You explain that value node is getting special treatment because Structured Configs assignment is this and that.
Maybe it's Structured Config assignment that is getting special treatment?

> We return when is a generic container because it can cause errors like omegaconf.errors.ConfigTypeError: Subscripted generics cannot be used with class and instance checks on ...

So instead of fixing the logic here, why can't you fix the check to handle Subscripted generics?

```python
validation_error = (
            target_type is not None
            and not is_generic_type(target_type)
            and value_type is not None
            and not issubclass(value_type, target_type)
        )
```

Also, does that mean that we can assign the value `1` to a subscripted generic?

Reply:
> Read this sentence again.
> You explain that value node is getting special treatment because Structured Configs assignment is this and that.
> Maybe it's Structured Config assignment that is getting special treatment?

Then how would you check for example:
```python
@dataclass
class DictClass:
    foo: Dict[str, int] = field(default_factory=lambda: {"a": 4})

@dataclass
class User:
    age: int = 4
    name: str = "foo"

cfg = OmegaConf.structured(DictClass)
cfg.foo = User() # should raise error because of name
```
It assigns a User which is a structured config but, since it doesn't have a ref_type to check it has to be checked when setting its values age and name. Am I missing something?

Structured Config is getting special treatment indirectly checking nodes. 
> > We return when is a generic container because it can cause errors like omegaconf.errors.ConfigTypeError: Subscripted generics cannot be used with class and instance checks on ...
> 
> So instead of fixing the logic here, why can't you fix the check to handle Subscripted generics?
> 
> ```python
> validation_error = (
>             target_type is not None
>             and not is_generic_type(target_type)
>             and value_type is not None
>             and not issubclass(value_type, target_type)
>         )
> ```
> 

I can move it to validation_error but isn't it logically the same? 

> Also, does that mean that we can assign the value `1` to a subscripted generic?

You can't, because 1 is being wrapped with a Dict annotation ref_type. It is tested on `test_assign_wrong_type_to_dict`